[pkg/ottl] Accept string (32 byte hex) {span,trace,profile} IDs

andrewvc opened the pull request:
#### Description

Currently `{TraceID,SpanID,ProfileID}` only support byte IDs, however in https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/43429 it was brought up that it is difficult to take a string representation of an ID and directly use that to set a trace ID. This would be for a hexadecimal string representation comprising 32 bytes. This change allows the `{TraceID,SpanID,ProfileID}` functions to also work on string inputs.

In short, the following now works:

```yaml
statements:
  - set(span.trace_id, TraceID("a389023abaa839283293ed323892389d"))
```

#### Link to tracking issue
Fixes https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/43429

#### Testing

In addition to the included go tests the following config 

[test-str-trace.yml](https://github.com/user-attachments/files/23217501/test-str-trace.yml)

was used to manually test by invoking `make otelcontribcol && ./bin/otelcontribcol_darwin_arm64  --config test-str-trace.yml` in one window and `telemetrygen traces --otlp-insecure --traces 1` in another. The output is as shown:
<details>
<summary>CLI Output</summary>

```
Span #0
    Trace ID       : a389023abaa839283293ed323892389d
    Parent ID      : e5c516b0c8942eab
    ID             : 1914136d5f9ca838
    Name           : okey-dokey-0
    Kind           : Server
    Start time     : 2025-10-29 18:59:03.312829 +0000 UTC
    End time       : 2025-10-29 18:59:03.312952 +0000 UTC
    Status code    : Unset
    Status message :
Attributes:
     -> network.peer.address: Str(1.2.3.4)
     -> peer.service: Str(telemetrygen-client)
Span #1
    Trace ID       : a389023abaa839283293ed323892389d
    Parent ID      :
    ID             : e5c516b0c8942eab
    Name           : lets-go
    Kind           : Client
    Start time     : 2025-10-29 18:59:03.312829 +0000 UTC
    End time       : 2025-10-29 18:59:03.312952 +0000 UTC
    Status code    : Unset
    Status message :
Attributes:
     -> network.peer.address: Str(1.2.3.4)
     -> peer.service: Str(telemetrygen-server)
```

</details>

#### Documentation

I don't currently see documentation for these functions, but I'm new here, maybe I'm missing something? Glad to update it

andrewvc committed:
Accept 32 char hex trace IDs

andrewvc committed:
Cleanup comment

andrewvc committed:
Add one last test

andrewvc committed:
Add changelog

andrewvc committed:
Fix linter errors

andrewvc commented:
I've added a fix for the linter issues, but it looks like perhaps an approval is required to run CI again?

andrewvc committed:
Fix chlog lint issues

andrewvc committed:
Merge remote-tracking branch 'origin/main' into accept-str-traceids

edmocosta commented on pkg/ottl/ottlfuncs/func_trace_id_test.go:
Let's also add an [e2e](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/pkg/ottl/e2e/e2e_test.go#L1163) test for that using string inputs.

andrewvc replied:
Done here! https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/43882/files#diff-90b6e27c60a80c3c831576a03b3bd0a47839bfb606a9d2876d8ccf6bf1370c92

edmocosta reviewed (commented):
Thank you @andrewvc and I'm sorry for the long delay to review. 
To keep it consistent, we also need to change the `SpanID` and `ProfileID` functions. If you agree, I'd be fine reviewing/including them on this same PR (changes are small enough), otherwise we can open separate PRs for that. Thanks!

andrewvc committed:
Merge remote-tracking branch 'origin/main' into accept-str-traceids

andrewvc commented on pkg/ottl/ottlfuncs/func_trace_id_test.go:
great idea

andrewvc committed:
Testing improvements from PR review

andrewvc committed:
Expand string support to profiles and spans

andrewvc committed:
Update chlog

andrewvc committed:
Merge branch 'main' into accept-str-traceids

andrewvc commented:
@edmocosta I've updated this PR and the PR description to mention all 3 functions per your request. This represents a significant refactor of the code to keep things DRY within reason, as well as accommodate 8 byte IDs

andrewvc committed:
Merge remote-tracking branch 'origin/main' into accept-str-traceids

andrewvc committed:
Add missing license headers

andrewvc committed:
Merge branch 'accept-str-traceids' of github.com:andrewvc/opentelemetry-collector-contrib into accept-str-traceids

andrewvc committed:
Update goporto links

andrewvc committed:
Fix broken test, fix linting

andrewvc committed:
Fix var shadowing for linter

andrewvc committed:
Incorporate PR feedback

- Add function name to error messages for ID parsing
- DRY up test helpers
- More precise types for id helper
- Un-export private helper

andrewvc commented on pkg/ottl/ottlfuncs/id_helpers.go:
```diff
@@ -0,0 +1,60 @@
+)
+
+type IDByteArray interface {
+	~[8]byte | ~[16]byte
```
I like this more precise approach, done!

andrewvc committed:
Extend function name testing to individual tests

andrewvc committed:
Merge remote-tracking branch 'origin/main' into accept-str-traceids

andrewvc committed:
Add comment to clarify test fn

andrewvc committed:
Update pkg/ottl/ottlfuncs/id_helpers.go

Co-authored-by: Edmo Vamerlatti Costa <11836452+edmocosta@users.noreply.github.com>

andrewvc committed:
Merge branch 'main' into accept-str-traceids

andrewvc committed:
Simplify error logic per PR feedback

andrewvc committed:
Merge remote-tracking branch 'andrewvc/accept-str-traceids' into accept-str-traceids

andrewvc commented on pkg/ottl/ottlfuncs/id_helpers.go:
```diff
@@ -0,0 +1,84 @@
+)
+
+// funcErrorType is an error type that includes the function name that caused the error.
+type funcErrorType struct {
```
I've tried to split the difference in https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/43882/commits/1926df323cba03cc25cabf78dc9bb49f4546d4cf

andrewvc committed:
Merge remote-tracking branch 'origin/main' into accept-str-traceids

andrewvc committed:
Add benchmarks for trace ID performance

andrewvc committed:
Add support for literal getters

andrewvc commented on pkg/ottl/ottlfuncs/func_profile_id.go:
```diff
@@ -4,21 +4,21 @@
+
 type ProfileIDArguments[K any] struct {
-	Bytes []byte
+	Target ottl.ByteSliceLikeGetter[K]
```
I think it just makes sense to use `ottl.GetLiteralValue` so I've added it to this branch. I've benchmarked it against main (and added benchmarks code to this PR in 7c19b76 ). You can view the results here: https://gist.github.com/andrewvc/4b9c73446cc1f9592177d3b980864c3a

andrewvc replied:
OK, new benchmark results (now they're just below). I had to fix a failing test by changing a type, and got a surprising result.

I had claude compare the two code bases and analyze why. 

FWIW I confirmed that having `main` also use `any` had the same result. The same change to `any` caused a similar speedup.

# Why Returning `any` Makes Literal TraceID 5x Faster

## The Performance Puzzle

After implementing literal optimization for the `TraceID` function, we observed surprising benchmark results:

| Implementation | ns/op | B/op | allocs/op |
|----------------|------:|-----:|----------:|
| Original (main) | 9.0 | 16 | 1 |
| New (literals) | 1.9 | 0 | 0 |

How did we achieve **5x faster** execution with **zero allocations**?

## Background: Interface Boxing in Go

When you assign a concrete type to an `any` (or any interface), Go performs "boxing" - it creates a small data structure containing:
1. A pointer to the type information
2. The value itself (or a pointer to it if too large)

For `pcommon.TraceID` (which is `[16]byte`), this boxing allocates 16 bytes on the heap.

## The Original Implementation

```go
func traceID[K any](bytes []byte) (ottl.ExprFunc[K], error) {
    // ... validation ...
    var idArr [16]byte
    copy(idArr[:16], bytes)
    id := pcommon.TraceID(idArr)  // id is concrete type [16]byte
    
    return func(context.Context, K) (any, error) {
        return id, nil  // ← Boxing happens HERE on every call
    }, nil
}
```

**The problem**: Every time the closure is called, `id` (a `[16]byte` array) must be boxed into the `any` return type. This boxing:
- Allocates 16 bytes
- Happens on **every single call**

## The New Implementation (Before Optimization)

Initially, our refactored code had the same issue:

```go
func bytesToID[R idByteArray](funcName string, b []byte, idLen, idHexLen int) (R, error) {
    var id R
    // ... populate id ...
    return id, nil  // Returns concrete type R
}

func newIDExprFunc[K any, R idByteArray](...) ottl.ExprFunc[K] {
    if b, ok := ottl.GetLiteralValue(target); ok {
        id, err := bytesToID[R](...)  // id is concrete type R
        return func(_ context.Context, _ K) (any, error) {
            return id, nil  // ← Still boxing on every call!
        }
    }
    // ...
}
```

## The Key Insight: Pre-Box at Construction Time

The fix was simple but profound - change `bytesToID` to return `any`:

```go
func bytesToID[R idByteArray](funcName string, b []byte, idLen, idHexLen int) (any, error) {
    var id R
    // ... populate id ...
    return id, nil  // ← Boxing happens HERE, once
}

func newIDExprFunc[K any, R idByteArray](...) ottl.ExprFunc[K] {
    if b, ok := ottl.GetLiteralValue(target); ok {
        result, err := bytesToID[R](...)  // result is already `any`
        return func(_ context.Context, _ K) (any, error) {
            return result, nil  // ← No boxing! Just returning existing `any`
        }
    }
    // ...
}
```

## Visual Comparison

### Before: Boxing Per Call
```
Construction:
  bytesToID() → [16]byte (concrete)
  closure captures: [16]byte

Every Call:
  closure() → box([16]byte) → any  ← ALLOCATION!
```

### After: Boxing Once
```
Construction:
  bytesToID() → any (already boxed) ← ALLOCATION (once)
  closure captures: any

Every Call:
  closure() → any  ← Just return existing value!
```

## Why This Works

1. **Interface values are just pointers**: Once a value is boxed into `any`, the interface value is essentially two pointers (type + data). Returning an `any` that's already boxed just copies these pointers.

2. **The closure captures the boxed value**: By capturing `result` (type `any`) instead of `id` (type `R`), the closure holds the already-boxed value.

3. **Type assertions are cheap**: When the caller does `result.(pcommon.TraceID)`, Go just checks the type pointer and returns the data - no allocation needed.

## The Numbers Don't Lie

**Branch:** `benchmark-original-traceid`, Commit: [`79925a1e77`](https://github.com/andrewvc/opentelemetry-collector-contrib/commit/79925a1e77)
```
BenchmarkTraceID/get_and_set-14    133809471    8.966 ns/op    16 B/op    1 allocs/op
```

**Branch:** `accept-str-traceids`, Commit: [`ca64679fbb`](https://github.com/andrewvc/opentelemetry-collector-contrib/commit/ca64679fbb)
```
BenchmarkTraceID/literal_bytes_get_and_set-14       659668681    1.785 ns/op    0 B/op    0 allocs/op
BenchmarkTraceID/literal_hex_string_get_and_set-14  669414519    1.798 ns/op    0 B/op    0 allocs/op
BenchmarkTraceID/dynamic_bytes_get_and_set-14       40554184    28.96 ns/op    40 B/op    2 allocs/op
```

The 16 bytes and 1 allocation have completely disappeared from the hot path!

## Conclusion

By changing `bytesToID` to return `any` instead of the concrete type `R`, we moved the boxing cost from **per-call** to **construction-time**. For literal values (the common case), this means:

- **5x faster execution** (9ns → 2ns)
- **Zero allocations per call** (1 → 0)
- **No additional memory traffic** in the hot path

This optimization applies to all three ID functions: `TraceID`, `SpanID`, and `ProfileID`. The literal optimization using `GetLiteralValue` ensures existing users get better performance, while the new hex string support comes at zero additional cost.

andrewvc committed:
Fix failing tests

andrewvc committed:
Benchmark get and set

andrewvc committed:
Add dynamic hex get and set

andrewvc committed:
Specialize hex decoding to reduce allocs for hex string decode

andrewvc committed:
Error early on init, ensure tests distinguish between init and dynamic failures

andrewvc committed:
Remove redundant test

andrewvc committed:
Merge remote-tracking branch 'origin/main' into accept-str-traceids

andrewvc committed:
Fix e2e tests

andrewvc committed:
Merge remote-tracking branch 'origin/main' into accept-str-traceids

andrewvc committed:
Remove unneeded makeLiteralIDGetter and helper

edmocosta commented on pkg/ottl/ottlfuncs/func_trace_id_test.go:
```diff
@@ -58,7 +60,8 @@ func BenchmarkTraceID(b *testing.B) {
 	b.Run("literal_bytes_get_and_set", func(b *testing.B) {
 		literalBytes := []byte{1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16}
-		literalGetter := makeLiteralIDGetter(literalBytes)
+		literalGetter, _ := ottl.NewTestingLiteralGetter(true, makeIDGetter(literalBytes))
```
Adding `require.NoError(b, err)` would do the work and make any possible failure clearer than panicking while executing the function. It's ok for this specific one, but considering how easy it's for handling it, I would do that.

edmocosta commented on pkg/ottl/ottlfuncs/func_trace_id_test.go:
```diff
@@ -6,53 +6,135 @@ package ottlfuncs
+	})
+}
+
+func BenchmarkTraceID(b *testing.B) {
```
Just a nit, feel free to ignore, but all benchmarks here are sharing the same code with different inputs, which works fine with the scenarios-based /table-driven approach:


```go
	benches := []struct {
		name    string
		value   []byte
		literal bool
	}{
		...
	}

	for _, bb := range benches {
		b.Run(bb.name, func(b *testing.B) {
			literalGetter, err := ottl.NewTestingLiteralGetter(bb.literal, makeIDGetter(bb.value))
			require.NoError(b, err)
			expr, err := traceID(literalGetter)
			require.NoError(b, err)
			b.ReportAllocs()
			b.ResetTimer()
			for b.Loop() {
				result, err := expr(b.Context(), nil)
				require.NoError(b, err)
			}
		})
	}
```

edmocosta reviewed (approved):
Great work @andrewvc! thank you!
I've left a few final nits, otherwise LGTM!

andrewvc commented on pkg/ottl/ottlfuncs/func_trace_id_test.go:
```diff
@@ -6,53 +6,135 @@ package ottlfuncs
+			if err != nil {
+				b.Fatal(err)
+			}
+			span.SetTraceID(result.(pcommon.TraceID))
```
I set it because I wanted to compare its result to the previous alternative `set(span.trace_id.string, Substring(Hex(cache["new_sha"]), 0, 32))` which I had benchmarked elsewhere. I also wanted to make sure that the we had something closer to the full path.

I think the benchmark could have been a bit more micro, but I think it's good because it susses out if any weird extra conversions need to happen with the `Set`. I know there aren't any due to ottl expression's `any`, but I think it is still nice.

andrewvc committed:
Switched to table driven benchmark

andrewvc committed:
Fix import ordering

andrewvc committed:
Merge branch 'main' into accept-str-traceids
